File: src/main/java/de/fraunhofer/aisec/cpg/frontends/cpp/StatementHandler.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */

package de.fraunhofer.aisec.cpg.frontends.cpp;

import de.fraunhofer.aisec.cpg.frontends.Handler;
import de.fraunhofer.aisec.cpg.graph.BreakStatement;
import de.fraunhofer.aisec.cpg.graph.CaseStatement;
import de.fraunhofer.aisec.cpg.graph.CatchClause;
import de.fraunhofer.aisec.cpg.graph.CompoundStatement;
import de.fraunhofer.aisec.cpg.graph.ContinueStatement;
import de.fraunhofer.aisec.cpg.graph.Declaration;
import de.fraunhofer.aisec.cpg.graph.DeclarationStatement;
import de.fraunhofer.aisec.cpg.graph.DefaultStatement;
import de.fraunhofer.aisec.cpg.graph.DoStatement;
import de.fraunhofer.aisec.cpg.graph.EmptyStatement;
import de.fraunhofer.aisec.cpg.graph.Expression;
import de.fraunhofer.aisec.cpg.graph.ForEachStatement;
import de.fraunhofer.aisec.cpg.graph.ForStatement;
import de.fraunhofer.aisec.cpg.graph.GotoStatement;
import de.fraunhofer.aisec.cpg.graph.IfStatement;
import de.fraunhofer.aisec.cpg.graph.LabelStatement;
import de.fraunhofer.aisec.cpg.graph.Literal;
import de.fraunhofer.aisec.cpg.graph.NodeBuilder;
import de.fraunhofer.aisec.cpg.graph.Region;
import de.fraunhofer.aisec.cpg.graph.ReturnStatement;
import de.fraunhofer.aisec.cpg.graph.Statement;
import de.fraunhofer.aisec.cpg.graph.SwitchStatement;
import de.fraunhofer.aisec.cpg.graph.TryStatement;
import de.fraunhofer.aisec.cpg.graph.Type;
import de.fraunhofer.aisec.cpg.graph.VariableDeclaration;
import de.fraunhofer.aisec.cpg.graph.WhileStatement;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;
import org.eclipse.cdt.core.dom.ast.IASTStatement;
import org.eclipse.cdt.core.dom.ast.IBinding;
import org.eclipse.cdt.core.dom.ast.ILabel;
import org.eclipse.cdt.core.dom.ast.cpp.ICPPASTCatchHandler;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTASMDeclaration;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTBreakStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTCaseStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTCatchHandler;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTCompoundStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTContinueStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTDeclarationStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTDefaultStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTDoStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTExpressionStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTForStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTGotoStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTIfStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTLabelStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTNullStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTRangeBasedForStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTReturnStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTSwitchStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTTryBlockStatement;
import org.eclipse.cdt.internal.core.dom.parser.cpp.CPPASTWhileStatement;

class StatementHandler extends Handler<Statement, IASTStatement, CXXLanguageFrontend> {

  StatementHandler(CXXLanguageFrontend lang) {
    super(Statement::new, lang);

    map.put(
        CPPASTCompoundStatement.class,
        ctx -> handleCompoundStatement((CPPASTCompoundStatement) ctx));
    map.put(CPPASTReturnStatement.class, ctx -> handleReturnStatement((CPPASTReturnStatement) ctx));
    map.put(
        CPPASTDeclarationStatement.class,
        ctx -> handleDeclarationStatement((CPPASTDeclarationStatement) ctx));
    map.put(
        CPPASTExpressionStatement.class,
        ctx -> handleExpressionStatement((CPPASTExpressionStatement) ctx));
    map.put(CPPASTIfStatement.class, ctx -> handleIfStatement((CPPASTIfStatement) ctx));
    map.put(CPPASTWhileStatement.class, ctx -> handleWhileStatement((CPPASTWhileStatement) ctx));
    map.put(CPPASTDoStatement.class, ctx -> handleDoStatement((CPPASTDoStatement) ctx));
    map.put(CPPASTForStatement.class, ctx -> handleForStatement((CPPASTForStatement) ctx));
    map.put(
        CPPASTRangeBasedForStatement.class,
        ctx -> handleForEachStatement((CPPASTRangeBasedForStatement) ctx));
    map.put(
        CPPASTContinueStatement.class,
        ctx -> handleContinueStatement((CPPASTContinueStatement) ctx));
    map.put(CPPASTBreakStatement.class, ctx -> handleBreakStatement((CPPASTBreakStatement) ctx));

    map.put(CPPASTLabelStatement.class, ctx -> handleLabelStatement((CPPASTLabelStatement) ctx));
    map.put(CPPASTSwitchStatement.class, ctx -> handleSwitchStatement((CPPASTSwitchStatement) ctx));
    map.put(CPPASTCaseStatement.class, ctx -> handleCaseStatement((CPPASTCaseStatement) ctx));
    map.put(
        CPPASTDefaultStatement.class, ctx -> handleDefaultStatement((CPPASTDefaultStatement) ctx));
    map.put(CPPASTNullStatement.class, ctx -> handleEmptyStatement((CPPASTNullStatement) ctx));
    map.put(CPPASTGotoStatement.class, ctx -> handleGotoStatement((CPPASTGotoStatement) ctx));
    map.put(
        CPPASTTryBlockStatement.class,
        ctx -> handleTryBlockStatement((CPPASTTryBlockStatement) ctx));
    map.put(CPPASTCatchHandler.class, ctx -> handleCatchHandler((ICPPASTCatchHandler) ctx));
  }

  private EmptyStatement handleEmptyStatement(CPPASTNullStatement emptyStmt) {
    return NodeBuilder.newEmptyStatement(emptyStmt.getRawSignature());
  }

  private TryStatement handleTryBlockStatement(CPPASTTryBlockStatement tryStmt) {
    TryStatement tryStatement = NodeBuilder.newTryStatement(tryStmt.toString());
    lang.getScopeManager().enterScope(tryStatement);
    CompoundStatement statement = (CompoundStatement) handle(tryStmt.getTryBody());

    List<CatchClause> catchClauses =
        Arrays.stream(tryStmt.getCatchHandlers())
            .map(this::handleCatchHandler)
            .collect(Collectors.toList());
    tryStatement.setTryBlock(statement);
    tryStatement.setCatchClauses(catchClauses);
    lang.getScopeManager().leaveScope(tryStatement);

    return tryStatement;
  }

  private CatchClause handleCatchHandler(ICPPASTCatchHandler catchHandler) {
    CatchClause catchClause = NodeBuilder.newCatchClause(catchHandler.getRawSignature());
    lang.getScopeManager().enterScope(catchClause);
    Statement body = this.lang.getStatementHandler().handle(catchHandler.getCatchBody());
    Declaration decl = null;
    if (catchHandler.getDeclaration() != null) { // can be null for "catch(...)"
      decl = this.lang.getDeclarationHandler().handle(catchHandler.getDeclaration());
    }
    catchClause.setBody((CompoundStatement) body);
    if (decl != null) {
      catchClause.setParameter((VariableDeclaration) decl);
    }
    lang.getScopeManager().leaveScope(catchClause);
    return catchClause;
  }

  private IfStatement handleIfStatement(CPPASTIfStatement ctx) {
    IfStatement statement = NodeBuilder.newIfStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(statement);
    if (ctx.getInitializerStatement() != null) {
      statement.setInitializerStatement(handle(ctx.getInitializerStatement()));
    }
    if (ctx.getConditionDeclaration() != null) {
      statement.setConditionDeclaration(
          this.lang.getDeclarationHandler().handle(ctx.getConditionDeclaration()));
    }
    statement.setConstExpression(ctx.isConstexpr());
    if (ctx.getConditionExpression() != null)
      statement.setCondition(this.lang.getExpressionHandler().handle(ctx.getConditionExpression()));
    statement.setThenStatement(handle(ctx.getThenClause()));

    if (ctx.getElseClause() != null) {
      statement.setElseStatement(handle(ctx.getElseClause()));
    }
    lang.getScopeManager().leaveScope(statement);
    return statement;
  }

  private LabelStatement handleLabelStatement(CPPASTLabelStatement ctx) {
    LabelStatement statement = NodeBuilder.newLabelStatement(ctx.getRawSignature());

    statement.setSubStatement(handle(ctx.getNestedStatement()));
    statement.setLabel(ctx.getName().toString());

    return statement;
  }

  private GotoStatement handleGotoStatement(CPPASTGotoStatement ctx) {
    GotoStatement statement = NodeBuilder.newGotoStatement(ctx.getRawSignature());
    BiConsumer assigneTargetLabel = (from, to) -> statement.setTargetLabel((LabelStatement) to);
    IBinding b = null;
    try {
      b = ctx.getName().resolveBinding();
      if (b instanceof ILabel) {
        ILabel label = (ILabel) b;
        label.getLabelStatement();
        // If the bound AST node is/or was transformed into a CPG node the cpg node is bound to the
        // CPG goto statement
        lang.registerObjectListener(label.getLabelStatement(), assigneTargetLabel);
      }
    } catch (Exception e) {
      // If the Label AST node was could not be resolved, the matchign is done based on label names
      // of CPG nodes using the predicate listeners
      lang.registerPredicateListener(
          (from, to) ->
              to instanceof LabelStatement
                  && ((LabelStatement) to).getLabel().equals(statement.getLabelName()),
          assigneTargetLabel);
    }
    return statement;
  }

  private WhileStatement handleWhileStatement(CPPASTWhileStatement ctx) {
    WhileStatement statement = NodeBuilder.newWhileStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(statement);
    if (ctx.getConditionDeclaration() != null)
      statement.setConditionDeclaration(
          this.lang.getDeclarationHandler().handle(ctx.getConditionDeclaration()));
    if (ctx.getCondition() != null)
      statement.setCondition(this.lang.getExpressionHandler().handle(ctx.getCondition()));
    statement.setStatement(handle(ctx.getBody()));
    lang.getScopeManager().leaveScope(statement);
    return statement;
  }

  private DoStatement handleDoStatement(CPPASTDoStatement ctx) {
    DoStatement statement = NodeBuilder.newDoStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(statement);
    statement.setCondition(this.lang.getExpressionHandler().handle(ctx.getCondition()));
    statement.setStatement(handle(ctx.getBody()));
    lang.getScopeManager().leaveScope(statement);

    return statement;
  }

  private ForStatement handleForStatement(CPPASTForStatement ctx) {
    ForStatement statement = NodeBuilder.newForStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(statement);
    statement.setInitializerStatement(handle(ctx.getInitializerStatement()));
    if (ctx.getConditionDeclaration() != null)
      statement.setConditionDeclaration(
          this.lang.getDeclarationHandler().handle(ctx.getConditionDeclaration()));
    if (ctx.getConditionExpression() != null)
      statement.setCondition(this.lang.getExpressionHandler().handle(ctx.getConditionExpression()));

    // Adds true expression node where default empty condition evaluates to true, remove here and in
    // java StatementAnalyzer
    if (statement.getConditionDeclaration() == null && statement.getCondition() == null) {
      Literal literal = NodeBuilder.newLiteral(true, new Type("bool"), "true");
      literal.setRegion(new Region());
      statement.setCondition(literal);
    }
    if (ctx.getIterationExpression() != null)
      statement.setIterationExpression(
          this.lang.getExpressionHandler().handle(ctx.getIterationExpression()));
    statement.setStatement(handle(ctx.getBody()));
    lang.getScopeManager().leaveScope(statement);

    return statement;
  }

  private ForEachStatement handleForEachStatement(CPPASTRangeBasedForStatement ctx) {
    ForEachStatement statement = NodeBuilder.newForEachStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(statement);

    Declaration var = this.lang.getDeclarationHandler().handle(ctx.getDeclaration());
    Statement iterable = this.lang.getExpressionHandler().handle(ctx.getInitializerClause());

    statement.setVariable(var);
    statement.setIterable(iterable);
    statement.setStatement(handle(ctx.getBody()));
    lang.getScopeManager().leaveScope(statement);

    return statement;
  }

  private BreakStatement handleBreakStatement(CPPASTBreakStatement ctx) {
    return NodeBuilder.newBreakStatement(ctx.getRawSignature());
    // C++ has no labeled break
  }

  private ContinueStatement handleContinueStatement(CPPASTContinueStatement ctx) {
    return NodeBuilder.newContinueStatement(ctx.getRawSignature());
    // C++ has no labeled continue
  }

  private Expression handleExpressionStatement(CPPASTExpressionStatement ctx) {
    return this.lang.getExpressionHandler().handle(ctx.getExpression());
  }

  private DeclarationStatement handleDeclarationStatement(CPPASTDeclarationStatement ctx) {
    if (ctx.getDeclaration() instanceof CPPASTASMDeclaration) {
      return NodeBuilder.newASMDeclarationStatement(ctx.getRawSignature());
    } else {
      DeclarationStatement declarationStatement =
          NodeBuilder.newDeclarationStatement(ctx.getRawSignature());

      declarationStatement.setDeclarations(
          this.lang.getDeclarationListHandler().handle(ctx.getDeclaration()));

      return declarationStatement;
    }
  }

  private ReturnStatement handleReturnStatement(CPPASTReturnStatement ctx) {
    ReturnStatement returnStatement = NodeBuilder.newReturnStatement(ctx.getRawSignature());
    // parse the return value
    // Todo Handle ReturnArgument
    if (ctx.getReturnValue() != null)
      returnStatement.setReturnValue(this.lang.getExpressionHandler().handle(ctx.getReturnValue()));

    return returnStatement;
  }

  private CompoundStatement handleCompoundStatement(CPPASTCompoundStatement ctx) {
    CompoundStatement compoundStatement = NodeBuilder.newCompoundStatement(ctx.getRawSignature());

    lang.getScopeManager().enterScope(compoundStatement);

    for (IASTStatement statement : ctx.getStatements()) {
      compoundStatement.getStatements().add(handle(statement));
    }
    lang.getScopeManager().leaveScope(compoundStatement);
    return compoundStatement;
  }

  private SwitchStatement handleSwitchStatement(CPPASTSwitchStatement ctx) {
    SwitchStatement switchStatement = NodeBuilder.newSwitchStatement(ctx.getRawSignature());
    lang.getScopeManager().enterScope(switchStatement);
    if (ctx.getInitializerStatement() != null)
      switchStatement.setInitializerStatement(handle(ctx.getInitializerStatement()));
    if (ctx.getControllerDeclaration() != null)
      switchStatement.setSelectorDeclaration(
          this.lang.getDeclarationHandler().handle(ctx.getControllerDeclaration()));
    if (ctx.getControllerExpression() != null)
      switchStatement.setSelector(
          this.lang.getExpressionHandler().handle(ctx.getControllerExpression()));
    switchStatement.setStatement(handle(ctx.getBody()));
    lang.getScopeManager().leaveScope(switchStatement);
    return switchStatement;
  }

  private CaseStatement handleCaseStatement(CPPASTCaseStatement ctx) {
    CaseStatement caseStatement = NodeBuilder.newCaseStatement(ctx.getRawSignature());
    caseStatement.setCaseExpression(this.lang.getExpressionHandler().handle(ctx.getExpression()));
    return caseStatement;
  }

  private DefaultStatement handleDefaultStatement(CPPASTDefaultStatement ctx) {
    return NodeBuilder.newDefaultStatement(ctx.getRawSignature());
  }
}

File: cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/StatementHandler.kt
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.frontends.cxx

import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.declarations.Declaration
import de.fraunhofer.aisec.cpg.graph.declarations.DeclarationSequence
import de.fraunhofer.aisec.cpg.graph.declarations.Variable
import de.fraunhofer.aisec.cpg.graph.expressions.*
import de.fraunhofer.aisec.cpg.graph.expressions.Block
import de.fraunhofer.aisec.cpg.graph.expressions.Expression
import de.fraunhofer.aisec.cpg.graph.expressions.Literal
import de.fraunhofer.aisec.cpg.graph.expressions.ProblemExpression
import de.fraunhofer.aisec.cpg.helpers.Util
import java.util.*
import java.util.function.BiConsumer
import java.util.function.Supplier
import java.util.stream.Collectors
import org.eclipse.cdt.core.dom.ast.*
import org.eclipse.cdt.core.dom.ast.cpp.ICPPASTCatchHandler
import org.eclipse.cdt.internal.core.dom.parser.cpp.*

class StatementHandler(lang: CXXLanguageFrontend) :
    CXXHandler<Expression, IASTStatement>(Supplier(::ProblemExpression), lang) {

    override fun handleNode(node: IASTStatement): Expression {
        return when (node) {
            is IASTCompoundStatement -> handleCompoundStatement(node)
            is IASTReturnStatement -> handleReturn(node)
            is IASTDeclarationStatement -> handleDeclarationStatement(node)
            is IASTExpressionStatement -> handleExpressionStatement(node)
            is IASTIfStatement -> handleIf(node)
            is IASTWhileStatement -> handleWhile(node)
            is IASTDoStatement -> handleDo(node)
            is IASTForStatement -> handleFor(node)
            is IASTContinueStatement -> handleContinue(node)
            is IASTBreakStatement -> handleBreak(node)
            is IASTLabelStatement -> handleLabel(node)
            is IASTSwitchStatement -> handleSwitch(node)
            is IASTCaseStatement -> handleCase(node)
            is IASTDefaultStatement -> handleDefault(node)
            is IASTNullStatement -> handleEmpty(node)
            is IASTGotoStatement -> handleGoto(node)
            is IASTProblemStatement -> handleProblemStatement(node)
            is CPPASTRangeBasedForStatement -> handleForEach(node)
            is CPPASTTryBlockStatement -> handleTryBlockStatement(node)
            is CPPASTCatchHandler -> handleCatchHandler(node)
            else -> {
                handleNotSupported(node, node.javaClass.name)
            }
        }
    }

    private fun handleProblemStatement(problemStatement: IASTProblemStatement): ProblemExpression {
        Util.errorWithFileLocation(
            frontend,
            problemStatement,
            log,
            problemStatement.problem.message,
        )

        return newProblemExpression(problemStatement.problem.message)
    }

    private fun handleEmpty(nullStatement: IASTNullStatement): Empty {
        return newEmpty(rawNode = nullStatement)
    }

    private fun handleTryBlockStatement(tryBlockStatement: CPPASTTryBlockStatement): Try {
        return newTry(enterScope = true) { tryStatement ->
            val statement = handle(tryBlockStatement.tryBody) as Block?
            val catchClauses =
                Arrays.stream(tryBlockStatement.catchHandlers)
                    .map { handleCatchHandler(it) }
                    .collect(Collectors.toList())
            tryStatement.tryBlock = statement
            tryStatement.catchClauses = catchClauses
        }
    }

    private fun handleCatchHandler(catchHandler: ICPPASTCatchHandler): CatchClause {
        return newCatchClause(rawNode = catchHandler, enterScope = true) { catchClause ->
            val body = frontend.statementHandler.handle(catchHandler.catchBody)

            // TODO: can also be an 'unnamed' parameter. In this case we should not declare a
            // variable
            var decl: Declaration? = null
            if (catchHandler.declaration != null) { // can be null for "catch(...)"
                decl = frontend.declarationHandler.handle(catchHandler.declaration)
            }

            catchClause.body = body as? Block

            if (decl is Variable) {
                frontend.scopeManager.addDeclaration(decl)
                catchClause.parameter = decl
            }
        }
    }

    private fun handleIf(ctx: IASTIfStatement): IfElse {
        return newIfElse(rawNode = ctx, enterScope = true) { statement ->
            // We need some special treatment for C++ IfStatements
            if (ctx is CPPASTIfStatement) {
                if (ctx.initializerStatement != null) {
                    statement.initializerStatement = handle(ctx.initializerStatement)
                }
                if (ctx.conditionDeclaration != null) {
                    statement.conditionDeclaration =
                        frontend.declarationHandler.handle(ctx.conditionDeclaration)
                }

                statement.isConstExpression = ctx.isConstexpr
            }

            if (ctx.conditionExpression != null)
                statement.condition = frontend.expressionHandler.handle(ctx.conditionExpression)
            statement.thenStatement = handle(ctx.thenClause)
            if (ctx.elseClause != null) {
                statement.elseStatement = handle(ctx.elseClause)
            }
        }
    }

    private fun handleLabel(ctx: IASTLabelStatement): Label {
        return newLabel(rawNode = ctx) { statement ->
            statement.subStatement = handle(ctx.nestedStatement)
            statement.label = ctx.name.toString()
            statement.name = newName(name = ctx.name.toString())
        }
    }

    private fun handleGoto(ctx: IASTGotoStatement): Goto {
        return newGoto(rawNode = ctx) { statement ->
            val assigneeTargetLabel = BiConsumer { _: Any, to: Node ->
                statement.targetLabel = to as Label
                to.label?.let {
                    statement.labelName = it
                    statement.name = newName(it)
                }
            }
            val b: IBinding?
            try {
                b = ctx.name.resolveBinding()
                if (b is ILabel) {
                    // If the bound AST node is/or was transformed into a CPG node the cpg node is
                    // bound to the CPG goto statement
                    frontend.registerObjectListener(b.labelStatement, assigneeTargetLabel)
                }
            } catch (_: Exception) {
                // If the Label AST node could not be resolved, the matching is done based on label
                // names of CPG nodes using the predicate listeners
                frontend.registerPredicateListener(
                    { _, to -> (to is Label && to.label == statement.labelName) },
                    assigneeTargetLabel,
                )
            }
        }
    }

    private fun handleWhile(ctx: IASTWhileStatement): While {
        return newWhile(rawNode = ctx, enterScope = true) { statement ->
            // Special treatment for C++ while
            if (ctx is CPPASTWhileStatement && ctx.conditionDeclaration != null) {
                statement.conditionDeclaration =
                    frontend.declarationHandler.handle(ctx.conditionDeclaration)
            }

            if (ctx.condition != null) {
                statement.condition = frontend.expressionHandler.handle(ctx.condition)
            }

            statement.statement = handle(ctx.body)
        }
    }

    private fun handleDo(ctx: IASTDoStatement): DoWhile {
        return newDoWhile(rawNode = ctx, enterScope = true) { statement ->
            statement.condition = frontend.expressionHandler.handle(ctx.condition)
            statement.statement = handle(ctx.body)
        }
    }

    private fun handleFor(ctx: IASTForStatement): For {
        return newFor(rawNode = ctx, enterScope = true) { statement ->
            statement.initializerStatement = handle(ctx.initializerStatement)

            // Special treatment for C++ while
            if (ctx is CPPASTForStatement && ctx.conditionDeclaration != null) {
                statement.conditionDeclaration =
                    frontend.declarationHandler.handle(ctx.conditionDeclaration)
            }

            if (ctx.conditionExpression != null) {
                statement.condition = frontend.expressionHandler.handle(ctx.conditionExpression)
            }

            // Adds true expression node where default empty condition evaluates to true, remove
            // here and in java StatementAnalyzer
            if (statement.conditionDeclaration == null && statement.condition == null) {
                val literal: Literal<*> =
                    newLiteral(true, primitiveType("bool")).implicit(code = "true")
                statement.condition = literal
            }

            if (ctx.iterationExpression != null) {
                statement.iterationStatement =
                    frontend.expressionHandler.handle(ctx.iterationExpression)
            }

            statement.statement = handle(ctx.body)
        }
    }

    private fun handleForEach(ctx: CPPASTRangeBasedForStatement): ForEach {
        return newForEach(rawNode = ctx, enterScope = true) { statement ->
            val decl = frontend.declarationHandler.handle(ctx.declaration)
            val `var` = newDeclarationStatement()
            `var`.singleDeclaration = decl
            val iterable: Expression? = frontend.expressionHandler.handle(ctx.initializerClause)
            statement.variable = `var`
            statement.iterable = iterable
            statement.statement = handle(ctx.body)
        }
    }

    private fun handleBreak(ctx: IASTBreakStatement): Break {
        return newBreak(rawNode = ctx)
        // C++ has no labeled break
    }

    private fun handleContinue(ctx: IASTContinueStatement): Continue {
        return newContinue(rawNode = ctx)
        // C++ has no labeled continue
    }

    private fun handleExpressionStatement(ctx: IASTExpressionStatement): Expression {
        val expression =
            frontend.expressionHandler.handle(ctx.expression)?.codeAndLocationFromOtherRawNode(ctx)
                ?: ProblemExpression("could not parse expression in statement")

        return expression
    }

    private fun handleDeclarationStatement(ctx: IASTDeclarationStatement): Expression {
        return if (ctx.declaration is IASTASMDeclaration) {
            // TODO: Specify the contained language through a language node and find a way to run a
            //  frontend for sub-block if available
            newDistinctLanguageBlock(rawNode = ctx)
        } else {
            newDeclarationStatement(rawNode = ctx) { declarationStatement ->
                val declaration = frontend.declarationHandler.handle(ctx.declaration)
                val declarations =
                    if (declaration is DeclarationSequence) {
                        declaration.asMutableList()
                    } else {
                        listOfNotNull(declaration)
                    }
                declarations.forEach {
                    frontend.scopeManager.addDeclaration(it)
                    declarationStatement.addDeclaration(it)
                }
            }
        }
    }

    private fun handleReturn(ctx: IASTReturnStatement): Return {
        return newReturn(rawNode = ctx) { returnStatement ->
            // Parse the return value
            if (ctx.returnValue != null) {
                returnStatement.returnValue = frontend.expressionHandler.handle(ctx.returnValue)
            }
        }
    }

    private fun handleCompoundStatement(ctx: IASTCompoundStatement): Block {
        return newBlock(rawNode = ctx, enterScope = true) { block ->
            for (statement in ctx.statements) {
                val handled = handle(statement)
                if (handled != null) {
                    block.statements += handled
                }
            }
        }
    }

    private fun handleSwitch(ctx: IASTSwitchStatement): Switch {
        return newSwitch(rawNode = ctx, enterScope = true) { switchStatement ->
            // Special treatment for C++ switch
            if (ctx is CPPASTSwitchStatement) {
                if (ctx.initializerStatement != null) {
                    switchStatement.initializerStatement = handle(ctx.initializerStatement)
                }
                if (ctx.controllerDeclaration != null) {
                    switchStatement.selectorDeclaration =
                        frontend.declarationHandler.handle(ctx.controllerDeclaration)
                }
            }

            if (ctx.controllerExpression != null) {
                switchStatement.selector =
                    frontend.expressionHandler.handle(ctx.controllerExpression)
            }

            switchStatement.statement = handle(ctx.body)
        }
    }

    private fun handleCase(ctx: IASTCaseStatement): Case {
        return newCase(rawNode = ctx) { caseStatement ->
            caseStatement.caseExpression = frontend.expressionHandler.handle(ctx.expression)
        }
    }

    private fun handleDefault(ctx: IASTDefaultStatement): Default {
        return newDefault(rawNode = ctx)
    }
}
